services/src/users/postgres_userdb.rs

```diff
@@ -0,0 +1,175 @@
+use crate::error;
+use crate::error::Result;
+use crate::users::session::{Session, SessionToken};
+use crate::users::user::User;
+use crate::users::user::{UserCredentials, UserId, UserRegistration};
+use crate::users::userdb::UserDB;
+use crate::util::identifiers::Identifier;
+use crate::util::user_input::Validated;
+use async_trait::async_trait;
+use bb8_postgres::bb8::Pool;
+use bb8_postgres::tokio_postgres::error::SqlState;
+use bb8_postgres::tokio_postgres::NoTls;
+use bb8_postgres::PostgresConnectionManager;
+use snafu::ResultExt;
+
+struct PostgresUserDB {
+    conn_pool: Pool<PostgresConnectionManager<NoTls>>, // TODO: support Tls connection as well
+}
+
+impl PostgresUserDB {
+    async fn new(conn_pool: Pool<PostgresConnectionManager<NoTls>>) -> Result<Self> {
+        let a = Self { conn_pool };
+        a.update_schema().await?;
+        Ok(a)
+    }
+
+    async fn schema_version(&self) -> Result<i32> {
+        // TODO: handle if version table not exists
+        let conn = self.conn_pool.get().await.map_err(|_| error::Error::BB8)?; // TODO: error reporting
+
+        let stmt = match conn.prepare("SELECT version from version").await {
+            Ok(stmt) => stmt,
+            Err(e) => {
+                if let Some(code) = e.code() {
+                    if *code == SqlState::UNDEFINED_TABLE {
+                        // TODO: log
+                        println!("UserDB: Uninitialized schema");
```

Comment:
Doesn't Johannes log output already in his gdal source?

Reply:
couldn't find it

Reply:
Okay, can be a new task. But it should be `eprintln!`.

Reply:
why? it's not an error.

Reply:
https://en.wikipedia.org/wiki/Standard_streams

`stderr` is not only about errors but also about status and diagnostics. For instance, if you have a program that converts A to B, then you don't want to print the loading status to stdout since it is expected to be the result of the program.

Reply:
da6a0baf66b03f687446b12cf4e672b69836995d